### PR TITLE
[FLINK-28250][Connector/Kafka] Fix exactly-once Kafka Sink out of memory

### PR DESCRIPTION
**Pull request by charles-tan**
## What is the purpose of the change

The goal of this PR is to address a memory leak where the Kafka Producer does not get cleaned up after kafka commit is successful.

## Brief change log

[7304f9d](https://github.com/charles-tan/flink/commit/7304f9dd3c2c65dd985f5de831496b091b58b556): recycles producer after successful kafka commit by adding a finally block to the existing try/catch.
[3d68e4e](https://github.com/apache/flink/commit/3d68e4e004d6c958c16c92ce8b5e4969183987ae): adds unit test
[c5fead3](https://github.com/apache/flink/pull/20205/commits/c5fead358fb70596a6363c443953a9f9ab8db6cc): removes finally block so that we don't recycle producer in the case of `RetriableException`

## Verifying this change

Manually verified the change by running a 1 node cluster with 1 JobManagers and 1 TaskManagers, running the following job for several hours without any memory issues:
```
public class FlinkTest {
    public void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        KafkaSource<String> source = KafkaSource.<String>builder()
            .setBootstrapServers("localhost:9092")
            .setTopics("input")
            .setGroupId("my-group" + System.currentTimeMillis())
            .setStartingOffsets(OffsetsInitializer.earliest())
            .setValueOnlyDeserializer(new SimpleStringSchema())
            .build();
        DataStream<String> sourceStream = env.fromSource(
            source,
            WatermarkStrategy.forMonotonousTimestamps(), "Kafka Source");
        KafkaRecordSerializationSchema<String> serializer = KafkaRecordSerializationSchema.builder()
            .setValueSerializationSchema(new SimpleStringSchema())
            .setTopic("output")
            .build();
        Properties sinkProps = new Properties();
        sinkProps.put("[transaction.timeout.ms](http://transaction.timeout.ms/)", 180000);
        KafkaSink<String> sink = KafkaSink.<String>builder()
            .setBootstrapServers("localhost:9092")
            .setKafkaProducerConfig(sinkProps)
            .setRecordSerializer(serializer)
            .setDeliverGuarantee(DeliveryGuarantee.EXACTLY_ONCE)
            .build();
        sourceStream.sinkTo(sink);
        env.enableCheckpointing(10000);
        env.getCheckpointConfig().setCheckpointTimeout(60000);
        env.getCheckpointConfig().setCheckpointingMode(CheckpointingMode.EXACTLY_ONCE);
        env.getCheckpointConfig().setMinPauseBetweenCheckpoints(500L);
        env.getCheckpointConfig().setTolerableCheckpointFailureNumber(1);
        env.execute("tester");
    }
}
```
Took heap dumps using `jmap` tool and noticed the taskmanager memory was stable. Link to the relevant mailing list thread: https://lists.apache.org/thread/c86cd8qyqb6qxy639hkzbozkwv2qxk84.

*Heap dump analysis*
Before patch using Flink 1.15.0 (note similar behavior observed with 1.15.1-r):
~30min: Screen Shot 2022-06-23 at 8 23 25 PM
~90min: Screen Shot 2022-06-23 at 8 23 44 PM

After patch using Flink 1.15.1-r:
~30min: Screen Shot 2022-07-07 at 4 10 33 PM
~90min: Screen Shot 2022-07-07 at 4 10 51 PM

This change was also noted in the JIRA ticket and validation was included in the ticket description.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaCommitter.java [hunk 1]
```diff
@@ -112,6 +112,8 @@ public void commit(Collection<CommitRequest<KafkaCommittable>> requests)
                         e);
                 recyclable.ifPresent(Recyclable::close);
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaCommitter.java:113]: What about removing this line as the `recyclable` will finally be closed anyway?

**Review by PatrickRen — COMMENTED** [review]
Thanks for the patch @charles-tan and the detailed test report! I left some comments. Also I think we could add a unit test for `KafkaCommitter` to validate this change.

BTW could you leave a comment in the original [JIRA ticket](https://issues.apache.org/jira/browse/FLINK-28250) to announce that you've taken this issue? I'll assign the ticket to you on JIRA then.

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Review by PatrickRen — COMMENTED** [review]
Thanks for the update @charles-tan !

The CI failed on case `KafkaCommitterTest#testRetryCommittableOnRetriableError` so I revisited the code in `KafkaCommitter`:

https://github.com/apache/flink/blob/e48abd38e8f3d8b5a1e2b9b5a2dfdb399614c236/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaCommitter.java#L75-L78

If the commit fails with a retriable exception, the commit request will be retried by committer operator in later checkpoints. We shouldn't recycle the producer under this case, so we can't put the recycle logic in the finally clause. 

Also I think we miss a `request.signalFailedWithKnownReason(e)` in the UnknownProducerIdException handler. Could you add it into this PR?

**Review by PatrickRen — APPROVED** [review]
Thanks for the patch @charles-tan ! LGTM. Could you squash all commits together? I'll merge the PR then.

Also we need to backport the patch to 1.15. Could you help to create the backport PR?

**Comment by charles-tan** [comment]
> Thanks for the patch @charles-tan ! LGTM. Could you squash all commits together? I'll merge the PR then.
> 
> Also we need to backport the patch to 1.15. Could you help to create the backport PR?

@PatrickRen thanks for the review! I've squashed the commits and created a backport patch for 1.15 [here](https://github.com/apache/flink/pull/20266).

**Pull request merged**